From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Sun, 23 Feb 2025 01:04:29 +0300
Subject: [PATCH] Option to disable saving of snowball and firework


diff --git a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index af9cf917f20a34479e01b63e686c65b2c783cea1..b344e6d0217500846852fa753075f1d58d3d50d8 100644
--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -350,4 +350,14 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
         double d1 = entity.position().z - this.position().z;
         return DoubleDoubleImmutablePair.of(d, d1);
     }
+
+    // DivineMC start - Option to disable saving firework
+    @Override
+    public boolean shouldBeSaved() {
+        if (this.level().divineConfig.disableFireworkSaving) {
+            return false;
+        }
+        return super.shouldBeSaved();
+    }
+    // DivineMC end - Option to disable saving firework
 }
diff --git a/net/minecraft/world/entity/projectile/Snowball.java b/net/minecraft/world/entity/projectile/Snowball.java
index c0b423967d6361a5d8ef11a5855c025e31948fe5..1dd963aae166e059107aac1d585ccc3199afb66d 100644
--- a/net/minecraft/world/entity/projectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/Snowball.java
@@ -94,4 +94,14 @@ public class Snowball extends ThrowableItemProjectile {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
     }
+
+    // DivineMC start - Option to disable snowball saving
+    @Override
+    public boolean shouldBeSaved() {
+        if (this.level().divineConfig.disableSnowballSaving) {
+            return false;
+        }
+        return super.shouldBeSaved();
+    }
+    // DivineMC end - Option to disable snowball saving
 }
